flow/tasks/os/exec.go

```diff
@@ -2,129 +2,205 @@ package os
 
 import (
 	"bytes"
+	"time"
+	"context"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"strings"
+	"sync"
+	"syscall"
 
 	"cuelang.org/go/cue"
-
 	hofcontext "github.com/hofstadter-io/hof/flow/context"
 	"github.com/hofstadter-io/hof/lib/cuetils"
 )
 
-type Exec struct{}
+type Exec struct {
+	cmd    *exec.Cmd
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
 
 func NewExec(val cue.Value) (hofcontext.Runner, error) {
 	return &Exec{}, nil
 }
 
-func (T *Exec) Run(ctx *hofcontext.Context) (interface{}, error) {
+func (T *Exec) Run(hctx *hofcontext.Context) (interface{}, error) {
+	
+	v := hctx.Value
+
+	// Create a cancellable context for process management
+	goctx, cancel := context.WithCancel(context.Background())
+	T.cancel = cancel
+	// Add defer at start of Run method
+	defer func() {
+			if T.cancel != nil {
+					T.cancel()
+			}
+	}()
 
-	v := ctx.Value
-	var cmd *exec.Cmd
+	// Setup channels for process management
+	processComplete := make(chan struct{}, 1)
+	processError := make(chan error, 1)
 
+	// Setup return value with defaults
+	ret := map[string]interface{}{
+		"exitcode": -1,
+		"success":  false,
+	}
+
+	// Lock CUE access and extract config
 	init_schemas(v.Context())
-	// unify with schema
 	v = v.Unify(task_exec)
 	if v.Err() != nil {
 		return nil, cuetils.ExpandCueError(v.Err())
 	}
-	// TODO, rework how i/o works for exec
+	
+	// Extract all configuration under lock
+	hctx.CUELock.Lock()
+	cmds, dir, env, stdin, stdout, stderr, doExit, exterr := extractExecConfig(v)
+	hctx.CUELock.Unlock()
+
+	if exterr != nil {
+		return nil, fmt.Errorf("os.Exec field: value extraction failed: %w", exterr)
+	}
 
-	// todo, check failure modes, fill, not return error?
-	// (in all tasks, really)
+	// Create and configure command with goctx
+	T.cmd = exec.CommandContext(goctx, cmds[0], cmds[1:]...)
+	T.cmd.Dir = dir
+	T.cmd.Env = env
 
-	var stdout, stderr io.Writer
+	// Setup I/O
+	if stdin != nil {
+		T.cmd.Stdin = stdin
+	}
+	if stdout != nil {
+		T.cmd.Stdout = stdout
+	}
+	if stderr != nil {
+		T.cmd.Stderr = stderr
+	}
 
-	ferr := func() error {
```

Comment:
didn't like this style tbh. i'm more of a clear, linear, conditional function dispatch kinda guy.